### PR TITLE
MH-13543 Rework workflow conditions, add string data type

### PR DESCRIPTION
**Pull request by pmiddend**
This PR reworks the “execution conditions” (the `if="…"` attribute on workflow operations). In the following mostly backwards-compatible kind of way:

# Switch the parser

After variable placeholders like `${foo}` are replaced in the condition string, we are left with a string denoting a boolean expression, like `true AND (1 < 2)`. This is currently parsed using the critically acclaimed `functional` library.

This PR removes the dependency on `functional` and instead adds a new, separate Opencast module which uses ANTLR to parse and interpret those strings. ANTLR generates Java code from a special grammar file, which looks very similar to the grammar given in the official Opencast developer documentation.

The code for replacing the variable placeholders was moved to this new module as well, as it seemed a bit more fitting.

# Add strings as possible values

The original condition language only knew booleans and numbers, along with operators such as addition and comparison. This PR adds the ability to handle strings, using the same set of operators. String literals are enclosed in apostrophes, and variables that don’t “look like numbers or booleans” are replaced by these quoted literals. This way, a comparison like this is possible:

```xml
<operation if="${captureAgentVendor} == 'ACME Corporation'">
…
</operation>
```

Operations that mix strings and numbers are “mostly harmless”. For example, `3+'4' == '34'` actually produces a value of `true`.

Note that “stringlike-looking” variables are *only* quoted inside `if` attributes. Variables in other places in the workflow XML are replaced as before, without quotes, so everything should be backwards-compatible.

# Add default values for variable placeholders

Variable placeholders like `${foo}` were previously replaced by `false` if the corresponding variable wasn’t found, which made no sense in relational contexts like `${foo} < 3` (frankly, I didn’t even check what happens in this case). This PR adds the ability to specify a default value after a colon (the syntax is borrowed from bash): `${foo:2}`.

This is, of course, *not backwards-compatible* and was made under the assumption that people don’t define workflow variable names with a colon in them. If that’s not the case, please leave a comment here!

# Misc

While working on the `WorkflowServiceImpl`, I removed some methods that weren't used and made minor style adjustments.

# Attribution

This work is sponsored by SWITCH.

### CODE REVIEW

File: modules/workflow-condition-parser/src/main/antlr4/org/opencastproject/workflow/conditionparser/antlr/WorkflowCondition.g4 [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+atom : '(' relationOperand ')' | NUMBER | STRING | BOOL ;
+
+AND: 'AND' ;
+OR: 'OR' ;
+NOT: 'NOT' ;
+NUMBER: '-'? [0-9]+ ('.' [0-9]+)? ;
+COMPARISONOPERATOR: '<=' | '<' | '==' | '>' | '!=' | '>=' ;
+NUMERICALOPERATOR: '+' | '-' | '*' | '/' ;
+BOOL: 'true' | 'false' ;
+VAR: '${' ('A'..'Z' | 'a'..'z' | '0'..'9') '}' ;
```
Comment (staubesv) [inline modules/workflow-condition-parser/src/main/antlr4/org/opencastproject/workflow/conditionparser/antlr/WorkflowCondition.g4]: This seems to restrictive: At least the underscore characters needs to be supported, too, as several WOHs created variables containing underscores. Likely, the minus character is in use, too.

Comment (staubesv) [reply modules/workflow-condition-parser/src/main/antlr4/org/opencastproject/workflow/conditionparser/antlr/WorkflowCondition.g4]: Actually `VAR` seems not to be used in the grammar at all. Can you please remove it?

Comment (pmiddend) [reply modules/workflow-condition-parser/src/main/antlr4/org/opencastproject/workflow/conditionparser/antlr/WorkflowCondition.g4]: Ah, yes, thanks for the reminder. I was going to evaluate the variables at the very end at first, so I had to keep the variables. Now that turned out different, so the placeholder parsing can be removed.

**Comment by mliradelc** [comment]
Hi, I just downloaded PR branch and I tried, but opencast crashes at startup.

```
org.apache.felix.resolver.reason.ReasonException: Unable to resolve root: missing requirement [root] osgi.identity; osgi.identity=opencast-allinone; type=karaf.feature; version="[8.0.0.SNAPSHOT,8.0.0.SNAPSHOT]"; filter:="(&(osgi.identity=opencast-allinone)(type=karaf.feature)(version>=8.0.0.SNAPSHOT)(version<=8.0.0.SNAPSHOT))" [caused by: Unable to resolve opencast-allinone/8.0.0.SNAPSHOT: missing requirement [opencast-allinone/8.0.0.SNAPSHOT] osgi.identity; osgi.identity=opencast-sox-impl; type=osgi.bundle; version="[8.0.0.SNAPSHOT,8.0.0.SNAPSHOT]"; resolution:=mandatory [caused by: Unable to resolve opencast-sox-impl/8.0.0.SNAPSHOT: missing requirement [opencast-sox-impl/8.0.0.SNAPSHOT] osgi.wiring.package; filter:="(&(osgi.wiring.package=org.opencastproject.workspace.api)(version>=8.0.0)(!(version>=9.0.0)))" [caused by: Unable to resolve opencast-workspace-api/8.0.0.SNAPSHOT: missing requirement [opencast-workspace-api/8.0.0.SNAPSHOT] osgi.wiring.package; filter:="(&(osgi.wiring.package=org.opencastproject.storage)(version>=8.0.0)(!(version>=9.0.0)))" [caused by: Unable to resolve opencast-common/8.0.0.SNAPSHOT: missing requirement [opencast-common/8.0.0.SNAPSHOT] osgi.identity; osgi.identity="root#opencast-core-8.0.0.SNAPSHOT"; type=karaf.subsystem; version="[0,0.0.0]"; resolution:=mandatory [caused by: Unable to resolve root#opencast-core-8.0.0.SNAPSHOT: missing requirement [root#opencast-core-8.0.0.SNAPSHOT] osgi.identity; osgi.identity=opencast-core; type=karaf.feature; version="[8.0.0.SNAPSHOT,8.0.0.SNAPSHOT]" [caused by: Unable to resolve opencast-core/8.0.0.SNAPSHOT: missing requirement [opencast-core/8.0.0.SNAPSHOT] osgi.identity; osgi.identity=opencast-workflow-condition-parser; type=osgi.bundle; version="[8.0.0.SNAPSHOT,8.0.0.SNAPSHOT]"; resolution:=mandatory [caused by: Unable to resolve opencast-workflow-condition-parser/8.0.0.SNAPSHOT: missing requirement [opencast-workflow-condition-parser/8.0.0.SNAPSHOT] osgi.wiring.package; filter:="(&(osgi.wiring.package=org.antlr.v4.runtime)(version>=4.7.0)(!(version>=5.0.0)))"]]]]]]]
	at org.apache.felix.resolver.Candidates$MissingRequirementError.toException(Candidates.java:1343)
	at org.apache.felix.resolver.ResolverImpl.doResolve(ResolverImpl.java:392)
	at org.apache.felix.resolver.ResolverImpl.resolve(ResolverImpl.java:378)
	at org.apache.felix.resolver.ResolverImpl.resolve(ResolverImpl.java:332)
	at org.apache.karaf.features.internal.region.SubsystemResolver.resolve(SubsystemResolver.java:257)
	at org.apache.karaf.features.internal.service.Deployer.deploy(Deployer.java:388)
	at org.apache.karaf.features.internal.service.FeaturesServiceImpl.doProvision(FeaturesServiceImpl.java:1025)
	at org.apache.karaf.features.internal.service.FeaturesServiceImpl.lambda$doProvisionInThread$13(FeaturesServiceImpl.java:964)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)
Caused by: org.apache.felix.resolver.reason.ReasonException: Unable to resolve opencast-allinone/8.0.0.SNAPSHOT: missing requirement [opencast-allinone/8.0.0.SNAPSHOT] osgi.identity; osgi.identity=opencast-sox-impl; type=osgi.bundle; version="[8.0.0.SNAPSHOT,8.0.0.SNAPSHOT]"; resolution:=mandatory [caused by: Unable to resolve opencast-sox-impl/8.0.0.SNAPSHOT: missing requirement [opencast-sox-impl/8.0.0.SNAPSHOT] osgi.wiring.package; filter:="(&(osgi.wiring.package=org.opencastproject.workspace.api)(version>=8.0.0)(!(version>=9.0.0)))" [caused by: Unable to resolve opencast-workspace-api/8.0.0.SNAPSHOT: missing requirement [opencast-workspace-api/8.0.0.SNAPSHOT] osgi.wiring.package; filter:="(&(osgi.wiring.package=org.opencastproject.storage)(version>=8.0.0)(!(version>=9.0.0)))" [caused by: Unable to resolve opencast-common/8.0.0.SNAPSHOT: missing requirement [opencast-common/8.0.0.SNAPSHOT] osgi.identity; osgi.identity="root#opencast-core-8.0.0.SNAPSHOT"; type=karaf.subsystem; version="[0,0.0.0]"; resolution:=mandatory [caused by: Unable to resolve root#opencast-core-8.0.0.SNAPSHOT: missing requirement [root#opencast-core-8.0.0.SNAPSHOT] osgi.identity; osgi.identity=opencast-core; type=karaf.feature; version="[8.0.0.SNAPSHOT,8.0.0.SNAPSHOT]" [caused by: Unable to resolve opencast-core/8.0.0.SNAPSHOT: missing requirement [opencast-core/8.0.0.SNAPSHOT] osgi.identity; osgi.identity=opencast-workflow-condition-parser; type=osgi.bundle; version="[8.0.0.SNAPSHOT,8.0.0.SNAPSHOT]"; resolution:=mandatory [caused by: Unable to resolve opencast-workflow-condition-parser/8.0.0.SNAPSHOT: missing requirement [opencast-workflow-condition-parser/8.0.0.SNAPSHOT] osgi.wiring.package; filter:="(&(osgi.wiring.package=org.antlr.v4.runtime)(version>=4.7.0)(!(version>=5.0.0)))"]]]]]]
	at org.apache.felix.resolver.Candidates$MissingRequirementError.toException(Candidates.java:1343)
	... 12 more
Caused by: org.apache.felix.resolver.reason.ReasonException: Unable to resolve opencast-sox-impl/8.0.0.SNAPSHOT: missing requirement [opencast-sox-impl/8.0.0.SNAPSHOT] osgi.wiring.package; filter:="(&(osgi.wiring.package=org.opencastproject.workspace.api)(version>=8.0.0)(!(version>=9.0.0)))" [caused by: Unable to resolve opencast-workspace-api/8.0.0.SNAPSHOT: missing requirement [opencast-workspace-api/8.0.0.SNAPSHOT] osgi.wiring.package; filter:="(&(osgi.wiring.package=org.opencastproject.storage)(version>=8.0.0)(!(version>=9.0.0)))" [caused by: Unable to resolve opencast-common/8.0.0.SNAPSHOT: missing requirement [opencast-common/8.0.0.SNAPSHOT] osgi.identity; osgi.identity="root#opencast-core-8.0.0.SNAPSHOT"; type=karaf.subsystem; version="[0,0.0.0]"; resolution:=mandatory [caused by: Unable to resolve root#opencast-core-8.0.0.SNAPSHOT: missing requirement [root#opencast-core-8.0.0.SNAPSHOT] osgi.identity; osgi.identity=opencast-core; type=karaf.feature; version="[8.0.0.SNAPSHOT,8.0.0.SNAPSHOT]" [caused by: Unable to resolve opencast-core/8.0.0.SNAPSHOT: missing requirement [opencast-core/8.0.0.SNAPSHOT] osgi.identity; osgi.identity=opencast-workflow-condition-parser; type=osgi.bundle; version="[8.0.0.SNAPSHOT,8.0.0.SNAPSHOT]"; resolution:=mandatory [caused by: Unable to resolve opencast-workflow-condition-parser/8.0.0.SNAPSHOT: missing requirement [opencast-workflow-condition-parser/8.0.0.SNAPSHOT] osgi.wiring.package; filter:="(&(osgi.wiring.package=org.antlr.v4.runtime)(version>=4.7.0)(!(version>=5.0.0)))"]]]]]
	at org.apache.felix.resolver.Candidates$MissingRequirementError.toException(Candidates.java:1343)
	... 13 more
Caused by: org.apache.felix.resolver.reason.ReasonException: Unable to resolve opencast-workspace-api/8.0.0.SNAPSHOT: missing requirement [opencast-workspace-api/8.0.0.SNAPSHOT] osgi.wiring.package; filter:="(&(osgi.wiring.package=org.opencastproject.storage)(version>=8.0.0)(!(version>=9.0.0)))" [caused by: Unable to resolve opencast-common/8.0.0.SNAPSHOT: missing requirement [opencast-common/8.0.0.SNAPSHOT] osgi.identity; osgi.identity="root#opencast-core-8.0.0.SNAPSHOT"; type=karaf.subsystem; version="[0,0.0.0]"; resolution:=mandatory [caused by: Unable to resolve root#opencast-core-8.0.0.SNAPSHOT: missing requirement [root#opencast-core-8.0.0.SNAPSHOT] osgi.identity; osgi.identity=opencast-core; type=karaf.feature; version="[8.0.0.SNAPSHOT,8.0.0.SNAPSHOT]" [caused by: Unable to resolve opencast-core/8.0.0.SNAPSHOT: missing requirement [opencast-core/8.0.0.SNAPSHOT] osgi.identity; osgi.identity=opencast-workflow-condition-parser; type=osgi.bundle; version="[8.0.0.SNAPSHOT,8.0.0.SNAPSHOT]"; resolution:=mandatory [caused by: Unable to resolve opencast-workflow-condition-parser/8.0.0.SNAPSHOT: missing requirement [opencast-workflow-condition-parser/8.0.0.SNAPSHOT] osgi.wiring.package; filter:="(&(osgi.wiring.package=org.antlr.v4.runtime)(version>=4.7.0)(!(version>=5.0.0)))"]]]]
	at org.apache.felix.resolver.Candidates$MissingRequirementError.toException(Candidates.java:1343)
	... 14 more
Caused by: org.apache.felix.resolver.reason.ReasonException: Unable to resolve opencast-common/8.0.0.SNAPSHOT: missing requirement [opencast-common/8.0.0.SNAPSHOT] osgi.identity; osgi.identity="root#opencast-core-8.0.0.SNAPSHOT"; type=karaf.subsystem; version="[0,0.0.0]"; resolution:=mandatory [caused by: Unable to resolve root#opencast-core-8.0.0.SNAPSHOT: missing requirement [root#opencast-core-8.0.0.SNAPSHOT] osgi.identity; osgi.identity=opencast-core; type=karaf.feature; version="[8.0.0.SNAPSHOT,8.0.0.SNAPSHOT]" [caused by: Unable to resolve opencast-core/8.0.0.SNAPSHOT: missing requirement [opencast-core/8.0.0.SNAPSHOT] osgi.identity; osgi.identity=opencast-workflow-condition-parser; type=osgi.bundle; version="[8.0.0.SNAPSHOT,8.0.0.SNAPSHOT]"; resolution:=mandatory [caused by: Unable to resolve opencast-workflow-condition-parser/8.0.0.SNAPSHOT: missing requirement [opencast-workflow-condition-parser/8.0.0.SNAPSHOT] osgi.wiring.package; filter:="(&(osgi.wiring.package=org.antlr.v4.runtime)(version>=4.7.0)(!(version>=5.0.0)))"]]]
	at org.apache.felix.resolver.Candidates$MissingRequirementError.toException(Candidates.java:1343)
	... 15 more
Caused by: org.apache.felix.resolver.reason.ReasonException: Unable to resolve root#opencast-core-8.0.0.SNAPSHOT: missing requirement [root#opencast-core-8.0.0.SNAPSHOT] osgi.identity; osgi.identity=opencast-core; type=karaf.feature; version="[8.0.0.SNAPSHOT,8.0.0.SNAPSHOT]" [caused by: Unable to resolve opencast-core/8.0.0.SNAPSHOT: missing requirement [opencast-core/8.0.0.SNAPSHOT] osgi.identity; osgi.identity=opencast-workflow-condition-parser; type=osgi.bundle; version="[8.0.0.SNAPSHOT,8.0.0.SNAPSHOT]"; resolution:=mandatory [caused by: Unable to resolve opencast-workflow-condition-parser/8.0.0.SNAPSHOT: missing requirement [opencast-workflow-condition-parser/8.0.0.SNAPSHOT] osgi.wiring.package; filter:="(&(osgi.wiring.package=org.antlr.v4.runtime)(version>=4.7.0)(!(version>=5.0.0)))"]]
	at org.apache.felix.resolver.Candidates$MissingRequirementError.toException(Candidates.java:1343)
	... 16 more
Caused by: org.apache.felix.resolver.reason.ReasonException: Unable to resolve opencast-core/8.0.0.SNAPSHOT: missing requirement [opencast-core/8.0.0.SNAPSHOT] osgi.identity; osgi.identity=opencast-workflow-condition-parser; type=osgi.bundle; version="[8.0.0.SNAPSHOT,8.0.0.SNAPSHOT]"; resolution:=mandatory [caused by: Unable to resolve opencast-workflow-condition-parser/8.0.0.SNAPSHOT: missing requirement [opencast-workflow-condition-parser/8.0.0.SNAPSHOT] osgi.wiring.package; filter:="(&(osgi.wiring.package=org.antlr.v4.runtime)(version>=4.7.0)(!(version>=5.0.0)))"]
	at org.apache.felix.resolver.Candidates$MissingRequirementError.toException(Candidates.java:1343)
	... 17 more
Caused by: org.apache.felix.resolver.reason.ReasonException: Unable to resolve opencast-workflow-condition-parser/8.0.0.SNAPSHOT: missing requirement [opencast-workflow-condition-parser/8.0.0.SNAPSHOT] osgi.wiring.package; filter:="(&(osgi.wiring.package=org.antlr.v4.runtime)(version>=4.7.0)(!(version>=5.0.0)))"
	at org.apache.felix.resolver.Candidates$MissingRequirementError.toException(Candidates.java:1343)
	... 18 more
```

After seeing this error I deleted and I downloaded the develop branch again and build it without this PR, worked without problems.

**Comment by staubesv** [comment]
@pmiddend I could reproduce the problem described by @mliradelc and it is related to antlr.

I've found the following related output when doing a `mvn clean install -Pdev`:

```
[INFO] --- maven-bundle-plugin:3.5.0:bundle (default-bundle) @ opencast-workflow-condition-parser ---
[WARNING] Bundle org.opencastproject:opencast-workflow-condition-parser:bundle:8-SNAPSHOT : Export org.opencastproject.workflow.conditionparser,  has 1,  private references [org.opencastproject.workflow.conditionparser.antlr], 
```

Could you please have a look at this?

Adding 

```
 <plugin>
       <groupId>org.apache.maven.plugins</groupId>
       <artifactId>maven-dependency-plugin</artifactId>
   </plugin>
```

to `/modules/workflow-service-impl/pom.xml` additionally reveals:

```
[INFO] --- maven-dependency-plugin:3.1.1:analyze-only (analyze) @ opencast-workflow-service-impl ---
[WARNING] Used undeclared dependencies found:
[WARNING]    javax.ws.rs:javax.ws.rs-api:jar:2.1:compile
[WARNING]    org.antlr:antlr4-runtime:jar:4.7.2:compile
[WARNING] Unused declared dependencies found:
[WARNING]    javax.ws.rs:jsr311-api:jar:1.1.1:compile
[WARNING]    org.slf4j:slf4j-log4j12:jar:1.7.21:test

```

**Comment by pmiddend** [comment]
Thanks for trying it out Maxi! Unfortunately, OSGi got in the way. I fixed that, it should work now.

**Comment by mliradelc** [comment]
> Thanks for trying it out Maxi! Unfortunately, OSGi got in the way. I fixed that, it should work now.

No problem!, I just checked the last version and now works, also I made some tests and works very good.

Can you add the examples you give here in the PR description in the documentation like this and the other one with the addition?

```xml
<operation if="${captureAgentVendor} == 'ACME Corporation'">
…
</operation>
```

**Comment by pmiddend** [comment]
@mliradelc Thanks for the feedback, I added two more examples to the documentation.

**Comment by mliradelc** [comment]
For me its good, I backported to Version 7 without issues. Version 6 it was very difficult because the Workflow Service changed quite a lot between 6 to 7.

If there are no other issues, I will merge it at the end of this day.

**Pull request merged**